examples/core/Graphormer/train.py

```diff
@@ -0,0 +1,491 @@
+            max_degree=max_degree, embedding_dim=embedding_dim
+        )
+
+        # path encoder
+        self.path_encoder = PathEncoder(
+            max_len=multi_hop_max_dist,
+            feat_dim=edge_dim,
+            num_heads=num_attention_heads,
+        )
+
+        # spatial encoder
+        self.spatial_encoder = SpatialEncoder(
+            max_dist=num_spatial, num_heads=num_attention_heads
+        )
+        self.graph_token_virtual_distance = nn.Embedding(1, num_attention_heads)
+
+        self.emb_layer_norm = nn.LayerNorm(self.embedding_dim)
+
+        self.layers = nn.ModuleList([])
+        self.layers.extend(
+            [
+                GraphormerLayer(
+                    feat_size=self.embedding_dim,
+                    hidden_size=ffn_embedding_dim,
+                    num_heads=num_attention_heads,
+                    dropout=dropout,
+                    activation=activation_fn,
+                    norm_first=pre_layernorm,
+                )
+                for _ in range(num_encoder_layers)
+            ]
+        )
+
+        # map graph_rep to num_classes
+        self.lm_head_transform_weight = nn.Linear(
+            self.embedding_dim, self.embedding_dim
+        )
+        self.layer_norm = nn.LayerNorm(self.embedding_dim)
+        self.activation_fn = activation_fn
+        self.embed_out = nn.Linear(self.embedding_dim, num_classes, bias=False)
+        self.lm_output_learned_bias = nn.Parameter(th.zeros(num_classes))
+
+    def reset_output_layer_parameters(self):
+        self.lm_output_learned_bias = nn.Parameter(th.zeros(1))
+        self.embed_out.reset_parameters()
+
+    def forward(
+        self,
+        node_feat,
+        in_degree,
+        out_degree,
+        path_data,
+        dist,
+        attn_mask=None,
+    ):
+        num_graphs, max_num_nodes, _ = node_feat.shape
+        deg_emb = self.degree_encoder(th.stack((in_degree, out_degree)))
+
+        # node feature + graph token
+        node_feat = self.atom_encoder(node_feat.int()).sum(dim=-2)
+        node_feat = node_feat + deg_emb
+        graph_token_feat = self.graph_token.weight.unsqueeze(0).repeat(
+            num_graphs, 1, 1
+        )
+        x = th.cat([graph_token_feat, node_feat], dim=1)
+
+        attn_bias = th.zeros(
+            num_graphs,
+            max_num_nodes + 1,
+            max_num_nodes + 1,
+            self.num_heads,
+            device=dist.device,
+        )
+        path_encoding = self.path_encoder(dist, path_data)
+        spatial_encoding = self.spatial_encoder(dist)
+        attn_bias[:, 1:, 1:, :] = path_encoding + spatial_encoding
+
+        # cls spatial encoder
+        t = self.graph_token_virtual_distance.weight.reshape(
+            1, 1, self.num_heads
+        )
+        # add spatial encoding between the virtual node and other nodes to attn_bias
+        attn_bias[:, 1:, 0, :] = attn_bias[:, 1:, 0, :] + t
+        attn_bias[:, 0, :, :] = attn_bias[:, 0, :, :] + t
```

Comment:
Why is there a `0` for L133 but not for L134?

Reply:
The shape of `attn_bias` is `[num_graphs, max_num_nodes, max_num_nodes, num_heads]`. Considering an easy situation with `num_graphs=1` and `num_heads=1`, where `attn_bias` can be regraded as an matrix with shape `[max_num_nodes, max_num_nodes]`. Since the virtual node comes first, the spatial encoding between it and other nodes needs to fill the 1st row (L134) and 1st column (L133) of `attn_bias`. Such spatial encoding is exactly the same for different node pairs and it will be broadcast to each position. To avoid `t` being added twice, we cannot replace `attn_bias[:, 1:, 0, :]` with `attn_bias[:, :, 0, :]` in L133.

Reply:
Got it. Thanks. Perhaps you can include a simplified version in the comments.